generic/tests/pktgen_perf.py

```diff
@@ -0,0 +1,197 @@
+
+    # get qemu, guest kernel and kvm version info and write them into result
+    result_path = utils_misc.get_path(test.resultsdir, "pktgen_perf.RHS")
+    result_file = open(result_path, "w")
+    kvm_ver = process.system_output(kvm_ver_chk_cmd)
+    host_ver = os.uname()[2]
+    result = session.cmd_output(guest_ver_cmd, timeout)
+    guest_ver = re.findall(".*?(\d{2}\.\d{2}\.\d{3}\.\d{4}).*?", result)
+    result_file.write("### kvm-userspace-ver : %s\n" % kvm_ver)
+    result_file.write("### kvm_version : %s\n" % host_ver)
+    result_file.write("### guest-kernel-ver :%s" % result)
+
+    # get record_list
+    record_line = ""
+    for record in record_list.split():
+        record_line += "%s|" % format_result(record)
+
+    # get result tested by each scenario
+    for pkt_cate in category.split():
+        result_file.write("Category:%s\n" % pkt_cate)
+        result_file.write("%s\n" % record_line.rstrip("|"))
+
+        # copy pktgen_test script to test server
+        local_path = os.path.join(data_dir.get_root_dir(),
+                                  "shared/scripts/pktgen_perf.sh")
+        remote_path = "/tmp/pktgen_perf.sh"
+        if pkt_cate == "tx":
+            pktgen_ip = vm.wait_for_get_address(0, timeout=5)
+            remote.scp_to_remote(pktgen_ip, shell_port, username,
+                                 password, local_path, remote_path)
+        elif pkt_cate == "rx":
+            process.run("cp %s %s" % (local_path, remote_path))
+
+        for size in pkt_size.split():
+            if pkt_cate == "tx":
+                error_context.context("test guest tx pps performance",
+                                      logging.info)
+                guest_mac = vm.get_mac_address(0)
+                pktgen_interface = utils_net.get_linux_ifname(session,
+                                                              guest_mac)
+                dsc_dev = utils_net.Interface(vm.get_ifname(0))
+                dsc = dsc_dev.get_mac()
+                runner = session.cmd
+                pkt_cate_r = run_test(session_serial, runner, remote_path,
+                                      pktgen_ip, dsc, pktgen_interface,
+                                      run_threads, size, timeout)
+            elif pkt_cate == "rx":
+                error_context.context("test guest rx pps performance",
+                                      logging.info)
+                host_bridge = params.get("netdst", "switch")
+                host_nic = utils_net.Interface(host_bridge)
+                pktgen_ip = host_nic.get_ip()
+                dsc = vm.wait_for_get_address(0, timeout=5)
+                pktgen_interface = vm.get_ifname(0)
+                runner = process.system_output
+                pkt_cate_r = run_test(session_serial, runner, remote_path,
+                                      pktgen_ip, dsc, pktgen_interface,
+                                      run_threads, size, timeout)
+            line = ""
```

Comment:
Unnecessary initialization?

Reply:
it must be empty, since need to separately record each small test's results in "for size in pkt_size.split()"

Here is a example result:
Category:tx
         pkt_size|        mpps
                   64|        0.37
Category:rx
         pkt_size|        mpps
                   64|        1.21

Reply:
@quanwenli I guess it does not matter here, each way you finally get same value for `line` as `"%s|%s" % (format_result(size), format_result(pkt_cate_r))`. You actually get the line separated by `\n` in `result_file.write(("%s\n" % line))` on line 143 below.

Reply:
Below errors is comment out "line = """
2018-03-15 02:37:52,302 stacktrace       L0047 ERROR|   File "/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/generic/tests/pktgen_perf.py", line 141, in run
2018-03-15 02:37:52,302 stacktrace       L0047 ERROR|     line += "%s|" % format_result(size)
2018-03-15 02:37:52,302 stacktrace       L0047 ERROR| UnboundLocalError: local variable 'line' referenced before assignment

Reply:
@quanwenli That's caused by `line += "%s|" % format_result(size)`, you should remove the `+` , which is `line = "%s|" % format_result(size)`

Reply:
yes, it make sense，I have update it. 

JOB LOG    : /root/avocado/job-results/job-2018-03-15T03.36-7509513/job.log
 (1/1) smp_8.4096m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.5.x86_64.io-github-autotest-qemu.pktgen_perf: PASS (265.94 s)